### DISCUSSION TITLE
Snapshotting / Rolling back Changes

### DISCUSSION THREAD

**Discussion by andykent** [groue/GRDB.swift#1311] [Q&A] [answered]
Summary: I am working on some data syncing algorithms. I would like a way to "rollback" the DB to a known previous point and then re-apply some changes. You could think of this as a fork and rebase. I am struggling to understand the best route to get this behaviour with GRDB / SQLite and any guidance would be appreciated.

Details:
- I need a way to create a checkpoint/snapshot of the DB at a given point in time. (a naive version would just be a copy of the DB file)
- I only ever intend there to be 1 of these snapshots, representing the state of the DB as it was after the last sync.
- The App will run off the mainline DB, not the snapshot so concurrency etc is not a huge issue.
- The snapshot will only ever be used as a checkpoint to rollback to so it doesn't need to be readable/writable beyond that, just a point in time backup.
- The App will keep it's own journal of changes of writes that have happened to the DB since the last snapshot and be responsible for replaying these along with handling conflicts etc.
- When a sync occurs the App will need a way to "rollback" to the last snapshot and then replay its journal of changes, ideally this should be possible atomically so that the UI only sees the end state.

Current thoughts:
- Whilst I am trying to avoid it the App _could_ maintain a rollback journal of every write so it can run them in reverse within a transaction to get back to the old state. This would be OK but it means manually keeping track of old data for deletes / updates and would be error prone.
- I could just copy the whole DB file but this seems less than ideal and would probably perform horribly. Also concurrency issues.
- I looked at `makeSnapshot()` this looks promising and functionally close to what I need but I would need to maintain that snapshot across App launches and a way to "rollback" to the snapshot.
- The SQLite Database Snapshot https://www.sqlite.org/c3ref/snapshot.html looks like it might do what I need but it's unclear to me if this is the same as a GRDB Snapshot (I think not?) and how I might go about making it work.
- I found this closed PR https://github.com/groue/GRDB.swift/pull/625 that looked like it might be getting closer to what I need, maybe.
- In my head this should be relatively easy if there is a WAL file as it's just a pointer into it and then a way to truncate the WAL at that point on demand. I am probably missing something important here though.

Is this even possible with SQLite / GRDB or should I give up and try doing to rollback journalling as an App concern? 
Any pointers or advice would be very welcome.

**Comment by groue** [comment, marked as answer]
Hello @andykent,

The word "snapshot" is full of different meanings, so let's first clarify what's a snapshot for SQLite (and GRBD). It is the same "snapshot" as in "snapshot isolation" ([Wikipedia definition](https://en.wikipedia.org/wiki/Snapshot_isolation), [SQLite definition](https://www.sqlite.org/isolation.html)).

It is the *behavior* exhibited by one particular runtime database connection, that starts when the connection enters snapshot isolation, and ends when the connection leaves snapshot isolation. During snapshot isolation, the connection sees an unchanging "snapshot" of the database file as it existed at the moment isolation was started. If other database connections perform concurrent writes, those writes are not seen by the isolated connection.

In SQLite, snapshot isolation starts with a read transaction, and ends when that transaction ends. In a [WAL SQLite database](https://www.sqlite.org/wal.html), snapshot isolation does not prevent concurrent writes. In other journaling modes, snapshot isolation is also guaranteed, but this is because concurrent writes are forbidden.

```sql
-- Enter snapshot isolation
BEGIN TRANSACTION;
-- Read count number 1.
SELECT COUNT(*) FROM player;
-- Read count number 2, guaranteed to be equal to count number 1.
SELECT COUNT(*) FROM player;
-- Leave snapshot isolation
COMMIT;

-- Enter snapshot isolation again
BEGIN TRANSACTION;
-- Read count number 3, not guaranteed to be equal to count number 1.
SELECT COUNT(*) FROM player;
COMMIT;
```

Snapshot isolation is a purely runtime concept, tied to one particular database connection, not to one particular database file. This means that when the database connection is closed, or when the process killed, this isolation ends, the "snapshot" vanishes.

What you are asking for is different, because you are looking for "snapshots" that outlive one particular runtime database connection. Those are more like backups, a word you have used as well.

> In my head this should be relatively easy if there is a WAL file as it's just a pointer into it and then a way to truncate the WAL at that point on demand. I am probably missing something important here though.

Yes. What you are missing is the lack of any SQLite api to define the "point in time" you want to refer to. Or rather, those apis exist, but such a "point in time" can only be referred to during the duration of a "snapshot isolation" (as described above), and are thus impossible to persist across process launches.

So. I hope I could help you understand that SQLite does not provide built-in support for the feature you are looking for (I mean, not according to my knowledge and understanding).

That said, copying the whole db file is one of your options indeed. You can choose between [`backup(to:)`](https://swiftpackageindex.com/groue/grdb.swift/documentation/grdb/databasereader/backup(to:pagesperstep:progress:)) and [`vacuum(into:)`](https://swiftpackageindex.com/groue/grdb.swift/documentation/grdb/databasewriter/vacuum(into:)-5lo41) - I'm not sure which one performs the best.

Doing smart tricks with the WAL file could be an option as well. As long as no checkpointing operation is performed, the main database file does not contain changes stored in the WAL file.  It is possible to prevent WAL checkpoints until you decide to define the next "backup" [^1]. Maybe SQLite could support going back in time if you just get rid of the WAL file? Take great care, though. There's no public SQLite api that tells "discard the whole WAL, I don't care about anything there". And check [How To Corrupt An SQLite Database File](https://www.sqlite.org/howtocorrupt.html) first.

[^1]: Beware: `PRAGMA wal_autocheckpoint` is not a persistent flag, so any write performed by some naive external connection may trigger a checkpoint and ruin your precious "backup".

**Reply by andykent** [reply]
Hello and thank you @groue! 

This is a super helpful reply, answered all my questions and helped me understand the snapshot capabilities better.

Whilst the end result isn't great news it has at least saved me from wasting a lot of time on the wrong path so thanks again.

I think I am just going to have to handle the rollback logic in Application code for now.

**Reply by groue** [reply]
Happy I could help, Andy! Happy GRDB, and happy 2023!
